File: LeetcodeCPP/501.cpp
#include <vector>
#include <queue>
#include <unordered_map>
#include <algorithm>
using namespace std;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<int> findMode(TreeNode* root) {
        unordered_map<int, int> node_counter;
        queue<TreeNode*> q;
        q.push(root);
        
        while (!q.empty()) {
            TreeNode* node = q.front();
            q.pop();
            node_counter[node->val]++;
            
            if (node->left) {
                q.push(node->left);
            }
            if (node->right) {
                q.push(node->right);
            }
        }
        
        int curr_mode = 0;
        vector<int> curr_list;
        
        for (const auto& pair : node_counter) {
            int key = pair.first;
            int val = pair.second;
            
            if (val > curr_mode) {
                curr_list = {key};
                curr_mode = val;
            } else if (val == curr_mode) {
                curr_list.push_back(key);
            }
        }
        
        return curr_list;
    }
};

File: LeetcodePython/501.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findMode(self, root: Optional[TreeNode]) -> List[int]:
        node_counter = defaultdict(int)
        q = deque()
        q.append(root)
        while q:
            node = q.popleft()
            node_counter[node.val] += 1
            if node.left:
                q.append(node.left)
            if node.right:
                q.append(node.right)
        curr_mode = 0
        curr_list = []
        for key, val in node_counter.items():
            if val > curr_mode:
                curr_list = [key]
                curr_mode = val
            elif val == curr_mode:
                curr_list.append(key)
        return curr_list
        